radix-engine/src/model/proof.rs

```diff
@@ -33,103 +45,314 @@ pub enum ProofError {
+                        if let Some(ids) = max.get_mut(source_id) {
+                            ids.extend(ids.clone());
+                        } else {
+                            max.insert(source_id.clone(), ids.clone());
+                        }
+                    }
+                }
+                Proof::Fungible { .. } => panic!("Illegal state"),
+            }
+        }
+
+        // check if the max satisfied the requested amount
+        let mut max_sum = BTreeSet::<NonFungibleId>::new();
+        for (_, value) in max {
+            max_sum.extend(value);
+        }
+        let total_ids = total_ids.unwrap_or(&max_sum);
+        if !max_sum.is_superset(&total_ids) {
+            return Err(ProofError::InsufficientBaseProofs);
+        }
+
+        // lock all relevant resources.
+        // See `compose_by_amount` for performance notes.
+        let mut new_sources = HashMap::new();
+        for proof in &proofs {
+            match proof {
+                Proof::NonFungible { sources, .. } => {
+                    for (source_id, (container, ids)) in sources {
+                        container
+                            .borrow_mut()
+                            .lock_ids(ids)
+                            .map_err(ProofError::ResourceContainerError)
+                            .expect("Re-locking should always succeed");
+                        new_sources.insert(source_id.clone(), (container.clone(), ids.clone()));
+                    }
+                }
+                Proof::Fungible { .. } => panic!("Illegal state"),
+            }
+        }
+
+        // issue a new proof
+        Proof::new_non_fungible(resource_def_id, false, total_ids.clone(), new_sources)
+    }
+
     pub fn clone(&self) -> Self {
-        for (container, amount_or_ids) in &self.locked_details {
-            match amount_or_ids {
-                AmountOrIds::Amount(amount) => container
-                    .borrow_mut()
-                    .lock_amount(amount.clone())
-                    .expect("Cloning should always be possible"),
-                AmountOrIds::Ids(ids) => container
-                    .borrow_mut()
-                    .lock_ids(ids)
-                    .expect("Cloning should always be possible"),
-            };
-        }
-
-        Self {
-            resource_def_id: self.resource_def_id,
-            resource_type: self.resource_type,
-            restricted: self.restricted,
-            locked_total: self.locked_total.clone(),
-            locked_details: self.locked_details.clone(),
+        match self {
+            Self::Fungible {
+                resource_def_id,
+                restricted,
+                total_amount,
+                sources,
+            } => {
+                for (container, amount) in sources.values() {
+                    container
+                        .borrow_mut()
+                        .lock_amount(amount.clone())
```

Comment:
State update in `clone()` is something worth mentioning in comments.

Reply:
Yes, will do